src/spec.c

```diff
@@ -1120,206 +1200,380 @@ static void IndexStats_RdbSave(RedisModuleIO *rdb, IndexStats *stats) {
-  //
-  //  // load the document, todo: use the key itself with scan api.
-  //  Document doc = {0};
-  //  Document_Init(&doc, keyname, 1.0, RS_LANG_ENGLISH);
-  //  if (Document_LoadAllFields(&doc, ctx) != REDISMODULE_OK) {
-  //    RedisModule_Log(ctx, "warning", "Failed loading document %*.s", (int)keynameCStrLen,
-  //                    keynameCStr);
-  //    Document_Free(&doc);
-  //    return;
-  //  }
-  //
-  //  QueryError status = {0};
-  //
-  //  dictIterator *iter = dictGetIterator(specDict);
-  //  dictEntry *entry = NULL;
-  //  while ((entry = dictNext(iter))) {
-  //    IndexSpec *sp = dictGetVal(entry);
-  //    if (IndexSpec_ShouldIndex(sp, keyname, key)) {
-  //
-  //      RedisSearchCtx sctx = SEARCH_CTX_STATIC(ctx, sp);
-  //
-  //      RSAddDocumentCtx *aCtx = NewAddDocumentCtx(sp, &doc, &status);
-  //      if (aCtx == NULL) {
-  //        // todo: handle this error
-  //        continue;
-  //      }
-  //
-  //      aCtx->donecb = IndexSpec_DoneIndexingCallabck;
-  //
-  //      aCtx->stateFlags |= ACTX_F_NOBLOCK;
-  //      aCtx->stateFlags |= ACTX_F_NOFREEDOC;
-  //
-  //      AddDocumentCtx_Submit(aCtx, &sctx, DOCUMENT_ADD_REPLACE);
-  //    }
-  //  }
-  //  dictReleaseIterator(iter);
-  //
-  //  // doc was set DEAD in Document_Moved and was not freed since it set as NOFREEDOC
-  //  doc.flags &= ~DOCUMENT_F_DEAD;
-  //  Document_Free(&doc);
+  IndexSpec *sp = scanner->spec_opt;
+  if (sp) {
+    IndexSpec_UpdateMatchingWithSchemaRules(sp, ctx, keyname);
+  } else {
+    Indexes_UpdateMatchingWithSchemaRules(ctx, keyname);
+  }
+  ++scanner->scannedKeys;
 }
 
-void IndexSpec_ScanAndReindexSpec(void *notused) {
+//---------------------------------------------------------------------------------------------
+
+#define REINDEX_BATCH_SIZE 100
+
+static void Indexes_ScanAndReindexTask(IndexesScanner *scanner) {
+  RS_LOG_ASSERT(scanner, "invalid IndexesScanner");
+
   RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(NULL);
   RedisModuleScanCursor *cursor = RedisModule_ScanCursorCreate();
-  //  RedisModule_ThreadSafeContextLock(ctx);
-  while (RedisModule_Scan(ctx, cursor, IndexSpec_ScanCallback, NULL)) {
-    //    RedisModule_ThreadSafeContextUnlock(ctx);
-    //    RedisModule_ThreadSafeContextLock(ctx);
+  RedisModule_ThreadSafeContextLock(ctx);
+
+  IndexSpec *sp = NULL;
+  if (scanner->spec_name_opt) {
+    sp = dictFetchValue(specDict, scanner->spec_name_opt);
+    if (!sp) { // spec was freed
+      goto end;
+    }
+    scanner->spec_opt = sp;
+  }
+
+  if (sp) {
+    RedisModule_Log(ctx, "notice", "Scanning index %s in background", sp->name);
+  } else {
+    RedisModule_Log(ctx, "notice", "Scanning indexes in background");
+  }
+
+  if (!sp) {
+    Indexes_SetTempSpecsTimers();
+  }
+
+  while (RedisModule_Scan(ctx, cursor, (RedisModuleScanCB) Indexes_ScanProc, scanner)) {
+    if (scanner->scannedKeys % REINDEX_BATCH_SIZE > 0) {
```

Comment:
You can not count on it, RedisModule_Scan might scan multiple and different amounts of keys each time. I guess its OK just to break any time it returns.

Reply:
The `scanner->scannedKeys` is really not the right number to check. I wanted to have some control over the yield-factor. We can count rounds or yield every time `RedisModule_Scan` returns.